config/7.1.0/securebanking/ig/scripts/groovy/com/forgerock/sapi/gateway/oauth/OAuthErrorResponseFactory.groovy

```diff
@@ -0,0 +1,170 @@
+        this.logPrefix = logPrefix
+        this.supportedMediaTypes = supportedMediaTypes
+        this.contentTypeNegotiator = new ContentTypeNegotiator(logPrefix, supportedMediaTypes)
+    }
+
+    Response invalidRequestErrorResponse(Header acceptHeader, Form errorForm) {
+        String bestContentType = contentTypeNegotiator.getBestContentType(acceptHeader)
+        errorForm.add("error", "invalid_request")
+        return errorResponse(Status.BAD_REQUEST, errorForm, bestContentType)
+    }
+
+     Response invalidClientErrorResponse(Header acceptHeader, Form errorForm) {
+         String bestContentType = contentTypeNegotiator.getBestContentType(acceptHeader)
+         errorForm.add("error", "invalid_client")
+         return errorResponse(Status.BAD_REQUEST, errorForm, bestContentType)
+     }
+
+     Response invalidGrantErrorResponse(Header acceptHeader, Form errorForm) {
+         String bestContentType = contentTypeNegotiator.getBestContentType(acceptHeader)
+         errorForm.add("error", "invalid_grant")
+         return errorResponse(Status.BAD_REQUEST, errorForm, bestContentType)
+     }
+
+     Response unsupportedGrantTypeErrorResponse(Header acceptHeader, Form errorForm) {
+         String bestContentType = contentTypeNegotiator.getBestContentType(acceptHeader)
+         errorForm.add("error", "invalid_grant")
+         return errorResponse(Status.BAD_REQUEST, errorForm, bestContentType)
+     }
+
+     Response unauthorizedClientErrorResponse(Header acceptHeader, Form errorForm) {
+         String bestContentType = contentTypeNegotiator.getBestContentType(acceptHeader)
+         errorForm.add("error", "unauthorized_client")
+         return errorResponse(Status.BAD_REQUEST, errorForm, bestContentType)
+     }
+
+     Response invalidScopeErrorResponse(Header acceptHeader, Form errorForm) {
+         String bestContentType = contentTypeNegotiator.getBestContentType(acceptHeader)
+         errorForm.add("error", "invalid_scope")
+         return errorResponse(Status.BAD_REQUEST, errorForm, bestContentType)
+     }
+
+    Response errorResponse(Status httpCode, Form errorForm, String bestContentType) {
+        String errorMessage = getErrorMessage(errorForm, bestContentType)
+        logger.warn("{} creating OAuth Error Response, http status: {}, error: {}", logPrefix, httpCode, errorMessage)
+        Response response = new Response(httpCode)
+        response.entity.setString(errorMessage)
+        ContentTypeHeader mediaTypeHeader = new ContentTypeHeader(bestContentType, [:])
+        response.addHeaders(mediaTypeHeader)
+        return response
+    }
+
+    String getErrorMessage(Form errorForm, String bestContentType) {
+        if(bestContentType == HttpMediaTypes.TEXT_HTML) {
+            return getHtmlErrorMessage(errorForm)
+        } else if (bestContentType == HttpMediaTypes.APPLICATION_FORM_URLENCODED) {
+            return getFormUrlEncodedErrorMessage(errorForm)
```

Comment:
Is this content type applicable for responses?

I would have assumed it was only applicable for requests i.e. it would be used to build the query string to append to requests to the redirect_uri to communicate an error response.

Unless I'm misunderstanding something, in this case it looks like it would be sent as the response entity as a string?

Reply:
Yes, it's there in case we need to append this to the redirect_uri - didn't take long to add, and if we ever do the jwt validation in the filter, then we can use the redirect_uri for errors and save ourselves a ton of screen capture uploads for FAPI compliance.